File: app/src/main/java/com/proxerme/app/util/listener/EndlessRecyclerOnScrollListener.java
package com.proxerme.app.util.listener;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;

import static android.support.v7.widget.RecyclerView.OnScrollListener;

/**
 * A listener for a RecyclerView, to load more items, once it reaches near the end.
 *
 * @author Ruben Gees
 */
public abstract class EndlessRecyclerOnScrollListener extends OnScrollListener {

    private static final int VISIBLE_THRESHOLD = 5;
    private int pastVisibleItems;
    private StaggeredGridLayoutManager mLayoutManager;

    public EndlessRecyclerOnScrollListener(@NonNull StaggeredGridLayoutManager mLayoutManager) {
        this.mLayoutManager = mLayoutManager;
    }

    @Override
    public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
        int visibleItemCount = mLayoutManager.getChildCount();
        int totalItemCount = mLayoutManager.getItemCount();
        int[] firstVisibleItems = new int[mLayoutManager.getSpanCount()];
        firstVisibleItems = mLayoutManager.findFirstVisibleItemPositions(firstVisibleItems);

        if (firstVisibleItems != null && firstVisibleItems.length > 0) {
            pastVisibleItems = firstVisibleItems[0];
        }

        if ((visibleItemCount + pastVisibleItems) >= (totalItemCount - VISIBLE_THRESHOLD)) {
            onLoadMore();
        }
    }

    public abstract void onLoadMore();
}

File: app/src/main/kotlin/com/proxerme/app/util/listener/EndlessRecyclerOnScrollListener.kt
package com.proxerme.app.util.listener

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.RecyclerView.OnScrollListener
import android.support.v7.widget.StaggeredGridLayoutManager

/**
 * A listener for a RecyclerView, to load more items, once it reaches near the end.

 * @author Ruben Gees
 */
abstract class EndlessRecyclerOnScrollListener(private val layoutManager:
                                               RecyclerView.LayoutManager) : OnScrollListener() {

    private companion object {
        const val VISIBLE_THRESHOLD = 5
    }

    private var lastScroll = 0L

    override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
        if (System.currentTimeMillis() - lastScroll > 50L) {
            val visibleItemCount = layoutManager.childCount
            val totalItemCount = layoutManager.itemCount
            var pastVisibleItems = 0

            if (layoutManager is StaggeredGridLayoutManager) {
                val firstVisibleItems = IntArray(layoutManager.spanCount)

                layoutManager.findFirstVisibleItemPositions(firstVisibleItems)

                if (firstVisibleItems.isNotEmpty()) {
                    pastVisibleItems = firstVisibleItems[0]
                }
            } else if (layoutManager is LinearLayoutManager) {
                pastVisibleItems = layoutManager.findFirstVisibleItemPosition()
            }

            if (totalItemCount > 0 && visibleItemCount + pastVisibleItems >=
                    totalItemCount - VISIBLE_THRESHOLD) {
                onLoadMore()
            }

            lastScroll = System.currentTimeMillis()
        }
    }

    abstract fun onLoadMore()
}